common/src/main/java/com/wynntils/gui/render/FontRenderer.java

```diff
@@ -82,6 +91,101 @@ public void renderText(
 
+    private void renderRainbowText(
+            PoseStack poseStack,
+            String text,
+            float x,
+            float y,
+            HorizontalAlignment horizontalAlignment,
+            VerticalAlignment verticalAlignment,
+            TextShadow shadow) {
+        float renderX;
+        float renderY;
+
+        renderX = switch (horizontalAlignment) {
+            case Left -> x;
+            case Center -> x - (font.width(text) / 2f);
+            case Right -> x - font.width(text);};
+
+        renderY = switch (verticalAlignment) {
+            case Top -> y;
+            case Middle -> y - (font.lineHeight / 2f);
+            case Bottom -> y - font.lineHeight;};
+
+        String formatting = "";
+
+        for (int i = 0; i < text.length(); i++) {
+            long timeDif = (long) renderX - 10;
+
+            String c = text.substring(i, i + 1);
+            if (c.equals("§")) {
+                if (!Objects.equals(getColorFromCode(text.charAt(i + 1)), "")) {
+                    formatting = "§" + text.charAt(i + 1);
+                } else {
+                    switch (text.charAt(i + 1)) {
+                        case 'l' -> formatting += "§l";
+                        case 'm' -> formatting += "§m";
+                        case 'n' -> formatting += "§n";
+                        case 'o' -> formatting += "§o";
+                        case 'r' -> formatting = "";
+                    }
+                }
+            }
+
+            // color settings
+            long time = System.currentTimeMillis() - timeDif;
+            float z = 2000.0F;
+
+            int colorInt = Color.HSBtoRGB((float) (time % (int) z) / z, 0.8F, 0.8F);
+            CustomColor color = CustomColor.fromInt(colorInt);
+
+            switch (shadow) {
+                case OUTLINE -> {
+                    int shadowColor = SHADOW_COLOR.withAlpha(color.a).asInt();
+                    String strippedText = ComponentUtils.stripColorFormatting(formatting + c);
+                    font.draw(poseStack, strippedText, renderX + 1, renderY, shadowColor);
+                    font.draw(poseStack, strippedText, renderX - 1, renderY, shadowColor);
+                    font.draw(poseStack, strippedText, renderX, renderY + 1, shadowColor);
+                    font.draw(poseStack, strippedText, renderX, renderY - 1, shadowColor);
+
+                    font.draw(poseStack, formatting + c, renderX, renderY, color.asInt());
+                }
+                case NORMAL -> {
+                    font.drawShadow(poseStack, formatting + c, renderX, renderY, color.asInt());
+                }
+                default -> {
+                    font.draw(poseStack, formatting + c, renderX, renderY, color.asInt());
+                }
+            }
+
+            // rendering the text
+            float charLength = font.width(c);
+            renderX += charLength + CHAR_SPACING;
+        }
+    }
+
+    public String getColorFromCode(char code) {
```

Comment:
Well this is certainly messy. I am not sure how to support chatformatting colors without weird code.

Reply:
At the very least there's likely a factory method on `Character` that can be used instead.